records.go

```diff
@@ -34,44 +38,75 @@ func (dht *IpfsDHT) GetPublicKey(ctx context.Context, p peer.ID) (ci.PubKey, err
 		return pk, nil
 	}
 
-	// ok, try the node itself. if they're overwhelmed or slow we can move on.
-	ctxT, cancelFunc := ctxfrac.WithDeadlineFraction(ctx, 0.3)
-	defer cancelFunc()
-	if pk, err := dht.getPublicKeyFromNode(ctx, p); err == nil {
-		err := dht.peerstore.AddPubKey(p, pk)
-		if err != nil {
-			return pk, err
+	// Try getting the public key both directly from the node it identifies
+	// and from the DHT, in parallel
+	resp := make(chan pubkrs, 2)
+	go func() {
```

Comment:
Need to use a cancelable context (so we can cancel one query if the other succeeds.